url56b1dc76d3_wufunxiang_Libs_tgz-pJ8-org_msgpack_util_TemplatePrecompilerJ8/src/org/msgpack/template/builder/ReflectionBeansTemplateBuilder.java

```diff
@@ -0,0 +1,210 @@
+//
+package org.msgpack.template.builder;
+
+import java.io.IOException;
+import java.lang.reflect.Method;
+import java.lang.reflect.Modifier;
+import java.lang.reflect.Type;
+import java.util.ArrayList;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+import org.msgpack.annotation.Ignore;
+import org.msgpack.annotation.Index;
+import org.msgpack.annotation.NotNullable;
+import org.msgpack.annotation.Optional;
+import org.msgpack.packer.Packer;
+import org.msgpack.template.FieldOption;
+import org.msgpack.template.Template;
+import org.msgpack.template.TemplateRegistry;
+import org.msgpack.template.builder.beans.BeanInfo;
+import org.msgpack.template.builder.beans.IntrospectionException;
+import org.msgpack.template.builder.beans.Introspector;
+import org.msgpack.template.builder.beans.PropertyDescriptor;
+import org.msgpack.unpacker.Unpacker;
+
+/**
+ * Class for building java reflection template builder for java beans class.
+ * 
+ * @author takeshita
+ * 
+ */
+@SuppressWarnings({ "rawtypes" })
+public class ReflectionBeansTemplateBuilder extends ReflectionTemplateBuilder {
+
+    private static Logger LOG = Logger.getLogger(ReflectionBeansTemplateBuilder.class.getName());
+
+    static class ReflectionBeansFieldTemplate extends ReflectionFieldTemplate {
+        ReflectionBeansFieldTemplate(final FieldEntry entry) {
+            super(entry);
+        }
+
+        
+        public void write(Packer packer, Object v, boolean required) throws IOException {
+            packer.write(v);
+        }
+
+        
+        public Object read(Unpacker unpacker, Object to, boolean required) throws IOException {
+            Object o = unpacker.read(entry.getType());
+            entry.set(to, o);
+            return o;
+        }
+    }
+
+    public ReflectionBeansTemplateBuilder(TemplateRegistry registry) {
+        super(registry, null);
+    }
+
+    
+    public boolean matchType(Type targetType, boolean hasAnnotation) {
+        Class<?> targetClass = (Class<?>) targetType;
+        boolean matched = matchAtBeansClassTemplateBuilder(targetClass, hasAnnotation);
+        if (matched && LOG.isLoggable(Level.FINE)) {
+            LOG.fine("matched type: " + targetClass.getName());
+        }
+        return matched;
+    }
+
+    
+    protected ReflectionFieldTemplate[] toTemplates(FieldEntry[] entries) {
+        ReflectionFieldTemplate[] tmpls = new ReflectionFieldTemplate[entries.length];
+        for (int i = 0; i < entries.length; i++) {
+            FieldEntry e = entries[i];
+            Class<?> type = e.getType();
+            if (type.isPrimitive()) {
+                tmpls[i] = new ReflectionBeansFieldTemplate(e);
+            } else {
+                Template tmpl = registry.lookup(e.getGenericType());
+                tmpls[i] = new FieldTemplateImpl(e, tmpl);
+            }
+        }
+        return tmpls;
+    }
+
+    
+    public FieldEntry[] toFieldEntries(Class<?> targetClass, FieldOption implicitOption) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 12. By comparison, 98% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 113-118 into a separate method.